src/vl-wizard-pane.js

```diff
@@ -0,0 +1,205 @@
+/**
+ * VlWizardPane
+ * @class
+ * @classdesc Gebruik de wizard pane om een stap in de wizard voor te stellen.
+ * 
+ * @extends VlElement
+ * 
+ * @property {boolean} data-vl-next-pane-disabled - Attribuut zorgt ervoor dat de gebruiker niet verder kan naar de volgende stap.
+ * @property {boolean} data-vl-previous-pane-disabled - Attribuut zorgt ervoor dat de gebruiker niet terug kan naar de vorige stap.
+ * 
+ * @see {@link https://www.github.com/milieuinfo/webcomponent-vl-ui-wizard/releases/latest|Release notes}
+ * @see {@link https://www.github.com/milieuinfo/webcomponent-vl-ui-wizard/issues|Issues}
+ * @see {@link https://webcomponenten.omgeving.vlaanderen.be/demo/vl-wizard.html|Demo}
+ * 
+ */
+export class VlWizardPane extends VlElement(HTMLElement) {
+    constructor() {
+        super(`
+            <style>
+                @import '/src/style.css';
+                @import '/node_modules/vl-ui-action-group/dist/style.css';
+                @import '/node_modules/vl-ui-grid/dist/style.css';
+                @import '/node_modules/vl-ui-button/dist/style.css';
+
+                :host {
+                    width: 100%;
+                }
+
+                slot[name="previous-action"], slot[name="next-action"] {
+                    display: inline-block;
+                }
+            </style>
+            <section class="vl-wizard__pane">
+                <slot name="title"></slot>
+                <div is="vl-grid" is-stacked>
+                    <div is="vl-column" size="12">
+                        <slot name="content"></slot>
+                    </div>
+                    <div is="vl-column" size="12">
+                        <div is="vl-action-group">
+                            <slot name="previous-action"></slot>
+                            <slot name="next-action"></slot>
+                        </div>
+                    </div>
+                </div>
+            </section>
+        `);
+        if (!this._previousActionSlot) {
+            this._shadow.querySelector('slot[name="previous-action"]').remove();
+        }
+        if (!this._nextActionSlot) {
+            this._shadow.querySelector('slot[name="next-action"]').remove();
+        }
+    }
+
+    connectedCallback() {
+        this._processTitle();
+        this._processActions();
+        this._observeActionsClick();
+    }
+
+    /**
+     * Geeft terug of de pagina actief is.
+     * 
+     * @return {Boolean}
+     */
+    get isActive() {
+        return this.classList.contains('is-selected') && !this.classList.contains('not-selected');
+    }
+
+    /**
+     * Geeft de titel van de pagina terug.
+     * 
+     * @return {String}
+     */
+    get title() {
+        return this._titleSlot ? this._titleSlot.innerText : undefined;
+    }
+
+    /**
+     * Geeft terug of de volgende pagina bereikt mag worden.
+     * 
+     * @return {Boolean}
+     */
+    get isNextPaneDisabled() {
+        return this.hasAttribute('data-vl-next-pane-disabled');
```

Comment:
panel moet niet beslissen wanneer het vorige of volgende beschikbaar is. panel moet zelf alleen aangeven wanneer het zelf compleet of disabled is, de rest beslist de wizard, wanneer er vorige en volgende en ... gedaan kan worden.

Reply:
dus voor mij moet al die enable en disable logica in de wizard zitten, das de enige die overzicht heeft op alle stappen en die dus bvb als er stappen zouden disabled worden kan beslissen er gerendered worden en zo, elke individuele stap heeft alleen maar als state nodig "disabled" en "complete" volgens mij

Reply:
anders heb je telkens heel veel werk met logica te veranderen als je eens wil 2 stappen van plaats wisselen ...

Reply:
Met de state 'disabled' en 'complete' kan je niet bepalen of vorige en/of volgende toegelaten zijn, alleen of de volgende toegelaten is. Momenteel kan de gebruiker dit met de attributen data-vl-next-pane-disabled en data-vl-previous-pane-disabled aangeven en wijzigen via de attributen of via de API. De opmerking doelt vooral op naamgeving lijkt me, dus als je betere namen hebt voor deze attributen/API functies zal ik ze toepassen.

Reply:
De huidige implementatie bij de demo pagina is als volgt:
```HTML
<vl-wizard id="vl-wizard-checkbox">
                <h2 is="vl-h2" slot="title">Wizard title</h2>
                <p slot="header">You're a wizard Harry</p>
                <div slot="panes">
                    <vl-wizard-pane id="vl-wizard-checkbox-pane-1" data-vl-next-pane-disabled>
                        <h3 is="vl-h3" slot="title">Stap 1</h3>
                        <div is="vl-form-grid" is-stacked slot="content">
                            <div is="vl-form-column" size="12">
                                <vl-checkbox id="vl-wizard-checkbox-pane-1-next" label="Volgende toegelaten"></vl-checkbox>
                            </div>
                        </div>
                        <button is="vl-button" type="button" slot="next-action">
                            Volgende
                        </button>
                    </vl-wizard-pane>
                    <vl-wizard-pane id="vl-wizard-checkbox-pane-2" data-vl-previous-pane-disabled data-vl-next-pane-disabled>
                        <h3 is="vl-h3" slot="title">Stap 2</h3>
                        <div is="vl-form-grid" is-stacked slot="content">
                            <div is="vl-form-column" size="12">
                                <vl-checkbox id="vl-wizard-checkbox-pane-2-previous" label="Vorige toegelaten"></vl-checkbox>
                                <vl-checkbox id="vl-wizard-checkbox-pane-2-next" label="Volgende toegelaten"></vl-checkbox>
                            </div>
                        </div>
                        <button is="vl-button-link" type="button" slot="previous-action">
                            <span is="vl-icon" data-vl-icon="arrow-left-fat" data-vl-before></span>
                            Vorige
                        </button>
                        <button is="vl-button" type="button" slot="next-action">
                            Volgende
                        </button>
                    </vl-wizard-pane>
                    <vl-wizard-pane id="vl-wizard-checkbox-pane-3" data-vl-previous-pane-disabled data-vl-next-pane-disabled>
                        <h3 is="vl-h3" slot="title">Stap 3</h3>
                        <div is="vl-form-grid" is-stacked slot="content">
                            <div is="vl-form-column" size="12">
                                <vl-checkbox id="vl-wizard-checkbox-pane-3-previous" label="Vorige toegelaten"></vl-checkbox>
                            </div>
                        </div>
                        <button is="vl-button-link" type="button" slot="previous-action">
                            <span is="vl-icon" data-vl-icon="arrow-left-fat" data-vl-before></span>
                            Vorige
                        </button>
                        <button is="vl-button" type="button" slot="next-action">
                            Versturen
                        </button>
                    </vl-wizard-pane>
                </div>
            </vl-wizard>
            <script>
                customElements.whenDefined('vl-wizard-pane').then(() => {
                    const pane1 = document.getElementById('vl-wizard-checkbox-pane-1');
                    const pane2 = document.getElementById('vl-wizard-checkbox-pane-2');
                    const pane3 = document.getElementById('vl-wizard-checkbox-pane-3');
                    document.getElementById('vl-wizard-checkbox-pane-1-next').oninput = event => event.detail ? pane1.enableNextPane() : pane1.disableNextPane();
                    document.getElementById('vl-wizard-checkbox-pane-2-previous').oninput = event => event.detail ? pane2.enablePreviousPane() : pane2.disablePreviousPane();
                    document.getElementById('vl-wizard-checkbox-pane-2-next').oninput = event => event.detail ? pane2.enableNextPane() : pane2.disableNextPane();
                    document.getElementById('vl-wizard-checkbox-pane-3-previous').oninput = event => event.detail ? pane3.enablePreviousPane() : pane3.disablePreviousPane();
                });
            </script>
        </div>
```

Misschien kan je besthier een voorbeeld van de code posten die je liever zou zien.

Reply:
Ik kan hier nu niet op 1-2-3 code uit mijn mouw schudden, maar het beheer van wanneer ge vorige of volgende kunt doen in een wizard hangt af van een combinatie van stappen, en dus van de wizard in zijn geheel en niet van 1 stap. 
Je kan in een wizard op volgende klikken als er na de huidige stap nog minstens 1 (non-disabled) stap volgt, en je kan in de wizard op vorige klikken als er nog minstens 1 (non-disabled) stap voor de huidige stap zit. En mogelijks wil je ook nog niet op volgende kunnen klikken als de huidige stap nog niet "complete" of "valid" of iets is.
De stappen van een wizard moeten op zich staande dingen zijn, waar je relatief makkelijk mee kan spelen (van plaats wisselen, disablen, ...) en dat gaat een veel moeilijker beheer vergen als je dat stap per stap (door de afnemer) laat beheren.

Reply:
Het is toch de stap die bepaalt of je naar de vorige/volgende stap mag los van de volgorde? Het is de wizard en de achterliggende code die de transitie op zich neemt, die logica zit niet in de stap. Wanneer je de volgorde van stappen wijzigt dan bepaalt de stap toch nog altijd of je naar de volgende/vorige stap mag gaan los van welke stap er voor of er achter komt? Ik zie ook geen usecase voor wijzigende volgorde in de stappen maar het zou op zich geen probleem geven met de huidige implementatie. Ik ben dus niet volledig mee met wat er anders moet, vandaar dat een code voorbeeld misschien iets kan verduidelijken?

Reply:
Nee, de stap bepaalt voor zich of hij zelf klaar is om naar de volgende over te gaan, of er nog vorige of volgende stappen zijn of niet daar weet een stap op zich niks van.

Reply:
Probeer maar eens uit te werken wat je zou moeten doen als je bvb een optionele stap hebt (die dus mag overgeslagen worden naargelang wat je in een vorige stap doet) en wat je dan moet doen als je met de volgorde van de stappen zou willen spelen.
Ik bedoel niet dat er een usecase is dat je on the fly stappen willen van volgorde veranderen, ik bedoel dat er een gewijzigde business vraag komt om stappen tussen te schuiven en optioneel te maken en zo.

Reply:
De stap weet inderdaad niet of er vorige of volgende stappen zijn, de transitie regelt de wizard. De usecase dat je beschrijft is mogelijk en kan ik toevoegen aan de demo pagina als je dat wilt.

Reply:
Ik vermoed dat je verward bent door de naamgeving van de functies.

Reply:
maar als de stap niet weet van andere stappen kan die toch ook niet weten of er nog een volgende knop moet zijn of niet.

Reply:
Nee, de afnemer bepaalt de knoppen per vl-wizard-pane en roept (eventueel) de functie(s) op. De functies kunnen ook opgeroepen worden wanneer er geen vorige/volgende knoppen zijn, ook al heeft dit weinig zin.

Reply:
het gaat mij puur om ease of development en minimaal aantal codewijzigingen om op een functionele change te reageren. die gaan met deze opzet veel groter zijn.

Reply:
Als je bvb 3 (onafhankelijke) stappen in je wizard zou hebben (A - B - C) en je zou die graag van plaats wisselen (B - C - A) vergt dat in jouw opzet wijzigingen aan elke stap zelf (vorige knop weg bij B, afronden knop vervangen door volgende knop bij C, volgende knop vervangen door afronden knop bij A), terwijl je eigenlijk gewoon code-gewijs A 2 lijntjes naar onder wil schuiven. 
Ik denk dat het dus voor mij blijft neerkomen op dat ik vind dat de knoppen dus niet per stap er moeten zijn, dus als dat onbespreekbaar blijft dan moet ik mij gewonnen geven denk ik ...

Reply:
Ik snap het argument dat de meerkost niet beperkt is tot het complexere gebruik, maar dat er ook naar onderhoudskost toe een impact is. Is default gedrag zoals pascal beschrijft dat overrulebaar is voor 'geavanceerder' gebruik zoals Tom voorzien heeft een optie? Indien ja, wat is de inspanning?

Reply:
Het gaat hier dus blijkbaar over #2. Ik denk niet dat de gebruiker één keer zijn stappen bepaalt, maar klopt inderdaad dat wanneer hij de 1e en laatste stap van plaats zou verwisselen hij inderdaad de vorige/volgende knop moet weg doen. Verder is er geen gevolg. #2 werd reeds ingeschat, dus @kevcraey aan jou om dit in te plannen of toe te voegen aan deze PR.

Reply:
Ik weet alleen ook niet of het handiger gaat zijn om in een complexe oplossing nog iets simpels aan toe te voegen of andersom ...

Reply:
aangezien er pros en cons zijn van beide oplossingen gaan we gebruiken door apocow en roxy afwachten om bij te sturen.